admin-dev/themes/new-theme/js/components/form/form-object-mapper.js

```diff
@@ -0,0 +1,173 @@
+/**
+ * Copyright since 2007 PrestaShop SA and Contributors
+ * PrestaShop is an International Registered Trademark & Property of PrestaShop SA
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.md.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+const {$} = window;
+
+/**
+ * This is able to watch an HTML form and parse it as a Javascript object based on a configurable
+ * mapping. Each field from the model is mapped to a form input, or several, each input is watched
+ * to keep the model consistent.
+ *
+ * When a model field has several inputs associated it keeps the in sync when either of them is updated.
+ */
+export default class FormObjectMapper {
+  /**
+   * @param $form {jQuery}
+   * @param modelMapping {Object}
+   * @param eventEmitter {EventEmitter}
+   * @param config {Object}
+   */
+  constructor($form, modelMapping, eventEmitter, config) {
+    this.$form = $form;
+    this.fullModelMapping = modelMapping;
+    this.eventEmitter = eventEmitter;
+
+    const inputConfig = config || {};
+
+    // This event is registered so when it is triggered it forces the object update
+    this.updateModelEventName = inputConfig.updateModel || 'updateModel';
+    // This event is emitted each time the object is updated (from both input change and external event)
+    this.modelUpdatedEventName = inputConfig.modelUpdated || 'modelUpdated';
+
+    this.modelMapping = {};
+    this.formMapping = {};
+    this.object = {};
+
+    this.initFormMapping();
+    this.watchUpdates();
+    this.updateFullObject();
+  }
+
+  getObject() {
+    return this.object;
+  }
+
+  /**
+   * Watches if changes happens from the form or via an event.
+   */
+  watchUpdates() {
```

Comment:
I don't know about you guys but it weirds me out not knowing a class method's visibility. I can't know what the class interface is. Why didn't we use underscores for private methods again?

Reply:
Using underscores won't make these members private, why would we try to use some principles the language is not made for ? Here is a good example why TypeScript would be good on the project (if you want to continue with classes), you would be able to use the TC39 Private Fields proposal like this : https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields or using private TS keyword

Reply:
So should I add `#` in front of each "private" method?

Reply:
Or add some `@private` tag? or something else? (until we use TypeScript)

Reply:
Except a comment, I don't think we can do anything to this :/

Reply:
We can do [private members](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields) with [Babel](https://babeljs.io/blog/2018/12/03/7.2.0)

Or we can go full on functional, or we can use TypeScript. Let's pick one and stop running in circles.

Reply:
Our babel version doesn't support private members

Reply:
We're not running in circles, it's not blocking anything 👀

Reply:
we could also return only public methods in class constructor as we've done with orders

Reply:
Choose one system and stick with it, either private members, pseudoprivate members or FP. But don't expose everything as public because otherwise we can't understand what the object's API is.

Reply:
Let's just return public methods in the constructor, this looks pretty fine to me :)

Really hope we'll adopt TS, this debate won't appear anymore